rotkehlchen/accounting/cost_basis.py

```diff
@@ -423,6 +465,115 @@ def calculate_spend_cost_basis(
             is_complete=is_complete,
         )
 
+    def calculate_spend_cost_basis_wac(
+            self,
+            spending_amount: FVal,
+            spending_asset: Asset,
+            timestamp: Timestamp,
+    ) -> CostBasisInfo:
+        """
+        When spending `spending_amount` of `spending_asset` at `timestamp` this function
+        calculates using the wac method the corresponding buy/s from
+        which to do profit calculation. Does NOT apply the "free after given time period"
+        rule which applies for some jurisdictions such as 1 year for Germany.
+
+        Returns the information in a CostBasisInfo object if enough acquisitions have
+        been found.
+        """
+        taxfree_bought_cost = ZERO
+        taxable_bought_cost = ZERO
+        taxable_amount = ZERO
+        is_complete = True
+        asset_events = self.get_events(spending_asset)
+
+        # for WAC, we need to rebuild the position up until timestamp. This is expensive.
+
+        # init a list of buy and sell we will use to build the pos
+        all_events = []
+
+        # add all buy that happened before timestamp
+        for acquisition_event in asset_events.acquisitions:
```

Comment:
This is heavy. For every single event you are re-calculating everything.

I think the better approach here would be to keep a rolling WAC result for each asset. So each even coming in adding to the previous average, and not having to recalculate the average for every new event.